### PR TITLE
Fix lost metadata when updating Model query

### LINKED ISSUE #22517: Adding/removing columns on an existing Model will drop all configured metadata
**Describe the bug**
Adding/removing columns (or any other action that causes a dirty-state) on an existing Model will drop all configured metadata.

**To Reproduce**
1. Question > SQL > Sample - save as "Q1"
```
select * from orders
```
2. Turn into a Model
3. Click "Customize metadata" and spend time on configuring that - Save changes
image
image
4. Click "Edit query definition" - change the query, so it adds/removes a column - e.g:
   ```
   select *, case when quantity>4 then 'large' else 'small' end size from orders
   ```
   image
5. Now all metadata has been lost (the 37% comes from auto-magic, which causes problems by itself [#19133](https://github.com/metabase/metabase/issues/19133))
image

**Expected behavior**
Don't drop metadata

**Information about your Metabase Installation:**
Tested 0.42.0 thru 0.43.0

### PR DESCRIPTION
**Pull request by gusaiani**
Fixes #22517 
### The problem
(Quoting @dpsutton comment [here](https://github.com/metabase/metabase/issues/22517#issuecomment-1143718139))
> So this bug happens because of the frontend. Editing the query doesn't actually erase the metadata. The problem happens when you run the query in the metadata editor. That hits a generic endpoint `POST api/dataset` with the edited query. This endpoint is not model related so doesn't use the model metadata. So the results end up without the edits you have in place. The bug is that the frontend keeps this metadata around instead of retaining the original metadata.
> 
> The metadata from this query run without a model context has only the "regular" metadata from the query itself. The frontend keeps this metadata associated with the model and sends it.
> 
> There are at least two possible ways to fix it:
> 
>   1. The frontend should just ignore this metadata from running the updated query. The backend will correctly handle this situation with the original edited metadata
> 
>   2. The frontend needs to send along the metadata when hitting the `POST api/dataset` when running the edited query. Either by enhancing `POST api/dataset` or a new endpoint that can handle this.
> 
> 
> Between these two, I think the second solution is better. Right now when previewing the edited query it does not show the effects of the edited metadata (column formatting, renaming, etc) so this feels like a bug. If we do the second option it allows this query preview to reflect the metadata changes.

### The solution

We went with the 2nd solution @dpsutton suggested
Which is:
- BE: updates `POST /api/dataset` to accept a new param `dataset-metadata`. This param is optional, when it's present we blend this `dataset-metadata` into the final results so that any user-edits metadata is preserved. 
  - We've done this by assoc the `dataset-metadata` as `[:info :metadata/dataset-metadata]` into our query. note that this key is what we use to preserve user-edits metadata for models.

- FE: If user is in a model, FE will now send the additional `dataset-metadata` to `POST /api/dataset`

### QP changes
This PR involves a change in the QP, in particular expanding the `resolve-fields` to resolve fields in `[:info :metadata/dataset-metdata]`.

To explain why I make this change, here is an example that will make QP fails to execute:
1. Checkout to this commit `579e491c62` -- this is where we implemented FE to send `dataset-metadata` but didn't fix the bug in QP.
2. Question > Orders > Save > Turn to model
3. Edit query definition > uncheck the created_at > Run

You'll see that the query will fail with an error: `Error: Field {created_at_id} is not present in the Query Processor Store.`

This is an error fired by `validate-temporal-bucketing` middleware and it fails because:
- The query is not selecting the `created_at` -- which is a DateTime column
  - So when we try to resolve fields in `resolve-fields` middleware, the `created_at` column **won't** be resolved
- But FE sends a `dataset-metadata` that contains the `created_at` column
  - And because the `validate-temporal-bucketting` find fields in the `outer-query` to validate, it'll find the `created_at` column inside `[:info :metadata/dataset-metadata]`
  - And because we haven't resolved it => qp.store throws an exception.

IMO this is just a symptom of the fact that we're not resolving fields inside `[:info :metadata/dataset-metadata]`, so I decided to update `resolve-fields` to resolves fields in it as well.

### How to Test

1. Question > SQL > Sample - save as "Q1"

```
select *, case when quantity>4 then 'large' else 'small' end size from orders
```

2. Turn into a Model
3. Click "Edit metadata" and spend time on configuring that, maybe rename column `ID`
4. Save changes
5. Click "Edit query definition" - change the query
   ```
   select *, case when quantity>4 then 'large' else 'smallyyyyyy' end size from orders
   ```
6. Run the query

Renamed columns should still be renamed after the query run.

### CODE REVIEW

**Comment by gusaiani** [comment]
For BE contributions, please take a look at this 2-min video:

https://www.loom.com/share/119cbb6ccfef4f8d89050f1e2ee18ad9

File: src/metabase/api/dataset.clj [hunk 1]
```diff
@@ -56,24 +59,35 @@
...
-                         (db/select-one [Card :result_metadata :dataset] :id source-card-id))
-        info           (cond-> {:executed-by api/*current-user-id*
-                                :context     context
-                                :card-id     source-card-id}
-                         (:dataset source-card)
-                         (assoc :metadata/dataset-metadata (:result_metadata source-card)))]
+  (let [source-card-id   (query->source-card-id query)
+        source-card      (when source-card-id
+                           (db/select-one [Card :result_metadata :dataset] :id source-card-id))
+        dataset-metadata (if (and (:dataset source-card) dataset-metadata)
```
Comment (qnkhuat) [inline src/metabase/api/dataset.clj:65]: If the source card is a dataset and dataset metadata is present, we'll combine them before passing it into `:info`.

**Review by daltojohnso — APPROVED** [review]
FE 👍

File: src/metabase/query_processor/middleware/resolve_fields.clj [hunk 2]
```diff
@@ -14,8 +14,10 @@
...
 (defn resolve-fields
   "Resolve all field referenced in the `query`, and store them in the QP Store."
-  [query]
-  (let [ids (into (set (mbql.u/match (:query query) [:field (id :guard integer?) _] id))
+  [{:keys [info] :as query}]
+  (let [ids (into (set (concat
+                         (mbql.u/match (:query query) [:field (id :guard integer?) _] id)
+                         (mbql.u/match (:metadata/dataset-metadata info) [:field (id :guard integer?) _] id)))
                   (comp cat (keep :id))
                   (mbql.u/match (:query query) {:source-metadata source-metadata} source-metadata))]
```
Comment (metamben) [inline src/metabase/query_processor/middleware/resolve_fields.clj:22]:
```suggestion
  (let [ids (-> (set (mbql.u/match (:query query) [:field (id :guard integer?) _] id))
                (into (mbql.u/match (:metadata/dataset-metadata info) [:field (id :guard integer?) _] id)
                (into (comp cat (keep :id))
                      (mbql.u/match (:query query) {:source-metadata source-metadata} source-metadata)))]
```

File: test/metabase/api/dataset_test.clj [hunk 3]
```diff
@@ -107,6 +108,44 @@
...
+        (testing "We shouldn't keep the dataset-metadata inside query when running it"
+          (let [run-query-async api.dataset/run-query-async]
+            (with-redefs [api.dataset/run-query-async (fn [query & opts]
+                                                        (assert (nil? (:dataset-metadata query)))
+                                                        (run-query-async query opts))]
+              (mt/user-http-request :rasta :post 202 "dataset" query))))))
+
+    (testing "should run successfully even if dataset-metadata contains fields that is not in the query"
+      (let [;; We're querying venues but this dataset-metadata contains a user-id
+            ;; This is to simulate the cases where `dataset-metadata` sent from FE can
```
Comment (metamben) [inline test/metabase/api/dataset_test.clj:136]:
```suggestion
            ;; This is to simulate the cases where `dataset-metadata` sent from FE
```

File: test/metabase/query_processor/middleware/resolve_fields_test.clj [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+(deftest resolves-field-in-metadata-dataset-metadata-test
+  ;; there are cases where FE can send a `dataset-metadata` via the POST /api/dataset
+  ;; and it could contains fields that might not included in the
+  ;; - selected-fields
+  ;; - source card
+  ;; - joined tables
+  ;; - etc
+  ;; We allow this behavior for editing models query, so let's make sure
+  ;; the `resolve-fields` find fields in [:info :metadata/dataset-metadata] as well.
+  (testing "`resolve-fields` should finds fields in [:info :metadata/dataset-metadata] (more context in #25000)"
```
Comment (metamben) [inline test/metabase/query_processor/middleware/resolve_fields_test.clj:21]:
```suggestion
  (testing "`resolve-fields` finds fields in [:info :metadata/dataset-metadata] (more context in #25000)"
```

File: test/metabase/query_processor/middleware/resolve_fields_test.clj [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+            [metabase.test :as mt]
+            [toucan.db :as db]))
+
+(defn- resolve-and-return-store-contents [query]
+  (mt/with-store-contents
+    (qp.resolve-fields/resolve-fields query)))
+
+(deftest resolves-field-in-metadata-dataset-metadata-test
+  ;; there are cases where FE can send a `dataset-metadata` via the POST /api/dataset
+  ;; and it could contains fields that might not included in the
```
Comment (metamben) [inline test/metabase/query_processor/middleware/resolve_fields_test.clj:14]:
```suggestion
  ;; containing fields that are not present in
```

File: test/metabase/query_processor/middleware/resolve_fields_test.clj [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  ;; - selected-fields
+  ;; - source card
+  ;; - joined tables
+  ;; - etc
+  ;; We allow this behavior for editing models query, so let's make sure
+  ;; the `resolve-fields` find fields in [:info :metadata/dataset-metadata] as well.
```
Comment (metamben) [inline test/metabase/query_processor/middleware/resolve_fields_test.clj:20]:
```suggestion
  ;; `resolve-fields` can find fields in [:info :metadata/dataset-metadata] as well.
```

**Review by metamben — APPROVED** [review]
BE looks good.

File: src/metabase/api/dataset.clj [hunk 7]
```diff
@@ -39,12 +40,14 @@
...
-(defn- run-query-async
+(defn run-query-async
+  "Run a query asynchronously."
   [{:keys [database], :as query}
-   & {:keys [context export-format qp-runner]
-      :or   {context       :ad-hoc
-             export-format :api
-             qp-runner     qp/process-query-and-save-with-max-results-constraints!}}]
+   & {:keys [dataset-metadata context export-format qp-runner]
+      :or   {dataset-metadata nil
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/api/dataset.clj:47]: It was just for documentation since all other params have default value.

File: src/metabase/api/dataset.clj [hunk 8]
```diff
@@ -56,24 +59,35 @@
...
-                                :context     context
-                                :card-id     source-card-id}
-                         (:dataset source-card)
-                         (assoc :metadata/dataset-metadata (:result_metadata source-card)))]
+  (let [source-card-id   (query->source-card-id query)
+        source-card      (when source-card-id
+                           (db/select-one [Card :result_metadata :dataset] :id source-card-id))
+        dataset-metadata (if (and (:dataset source-card) dataset-metadata)
+                           (qp.util/combine-metadata (:result_metadata source-card) dataset-metadata)
+                           (or (:result_metadata source-card) dataset-metadata nil))
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/api/dataset.clj:67]: I was just being explicit with what will be returned if none of those metadata are present. But on second thought I think it's not needed. I will remove it.

File: src/metabase/query_processor/middleware/resolve_fields.clj (hunk 2 above)
Comment (qnkhuat) [reply src/metabase/query_processor/middleware/resolve_fields.clj:22]: much more readable, thanks!

File: src/metabase/api/dataset.clj [hunk 9]
```diff
@@ -39,12 +40,13 @@
...
+  "Run a query asynchronously."
   [{:keys [database], :as query}
-   & {:keys [context export-format qp-runner]
-      :or   {context       :ad-hoc
-             export-format :api
-             qp-runner     qp/process-query-and-save-with-max-results-constraints!}}]
+   & {:keys [dataset-metadata context export-format qp-runner]
+      :or   {context          :ad-hoc
+             export-format    :api
+             qp-runner        qp/process-query-and-save-with-max-results-constraints!}}]
```
Comment (camsaul) [inline src/metabase/api/dataset.clj:49]: Did you mean to change the indentation like this?

File: test/metabase/api/dataset_test.clj [hunk 10]
```diff
@@ -107,6 +108,44 @@
...
+            ;; Since we're allowing this behavior of sending custom dataset-metadata
+            ;; we expect the query to run successfully.
+            dataset-metadata [{:name          "ID"
+                               :display_name  "USER ID CAPITALIZED"
+                               :description   "User id"
+                               :semantic_type "type/FK"
+                               :base_type     "type/BigInteger"
+                               :field_ref     ["field" (mt/id :users :id) nil]}]]
+        (mt/user-http-request :rasta :post 202 "dataset" (assoc (mt/mbql-query venues {:fields [$id]})
+                                                          :dataset-metadata dataset-metadata))))))
```
Comment (camsaul) [inline test/metabase/api/dataset_test.clj:147]: Shouldn't there be some sort of `is` assertion to check the results? Doesn't seem like this isn't really testing anything

File: test/metabase/api/dataset_test.clj [hunk 11]
```diff
@@ -107,6 +108,44 @@
...
+               (->> (mt/user-http-request :rasta :post 202 "dataset" query)
+                    mt/cols
+                    (map #(select-keys % [:name :display_name :description :base_type :field_ref :semantic_type])))))
+
+        (testing "We shouldn't keep the dataset-metadata inside query when running it"
+          (let [run-query-async api.dataset/run-query-async]
+            (with-redefs [api.dataset/run-query-async (fn [query & opts]
+                                                        (assert (nil? (:dataset-metadata query)))
+                                                        (run-query-async query opts))]
+              (mt/user-http-request :rasta :post 202 "dataset" query))))))
```
Comment (camsaul) [inline test/metabase/api/dataset_test.clj:132]: This is also not testing anything other than that the query runs... it could easily be running just fine for some other reason and completely ignore the metadata you're passing

File: src/metabase/query_processor/middleware/resolve_fields.clj [hunk 12]
```diff
@@ -14,10 +14,11 @@
...
   "Resolve all field referenced in the `query`, and store them in the QP Store."
-  [query]
-  (let [ids (into (set (mbql.u/match (:query query) [:field (id :guard integer?) _] id))
-                  (comp cat (keep :id))
-                  (mbql.u/match (:query query) {:source-metadata source-metadata} source-metadata))]
+  [{:keys [info] :as query}]
+  (let [ids (-> (set (mbql.u/match (:query query) [:field (id :guard integer?) _] id))
+                (into (mbql.u/match (:metadata/dataset-metadata info) [:field (id :guard integer?) _] id))
+                (into (comp cat (keep :id))
+                      (mbql.u/match (:query query) {:source-metadata source-metadata} source-metadata)))]
```
Comment (camsaul) [inline src/metabase/query_processor/middleware/resolve_fields.clj:21]: Usually instead of threading together `into` forms you can combine them into a single one using transducers and `comp`... since that was how it was in the first place I would suggest keeping it that way and just adding more transducers to the `comp`. I think that would be easier to read than threading together several `into`s

File: test/metabase/api/dataset_test.clj (hunk 11 above)
Comment (camsaul) [reply test/metabase/api/dataset_test.clj:132]: keep in mind because the QP endpoint is async it always returns 202 whether it succeeds or not... you need to look at the response

**Review by camsaul — CHANGES_REQUESTED** [review]
1. It looks like a lot of the tests aren't really testing anything. The QP endpoint is async and will return a 202 response even if the query fails. You need to also check the response and make sure the query actually succeeds, and returns the merged metadata you expect.

2. I'm not **convinced** this isn't going to cause potential security issues. Adding tests to prove that it won't will make me feel better.

   My main concern is that someone can abuse the stuff you added to resolve

   > a field that is totally unrelated to the query

   and potentially trigger an implicit join against a different table that they're not supposed to be able to see, or maybe filter values for a field they're not able to see. For example if I'm only allowed to see the test data `venues` table and I send `:metadata/dataset-metadata` for `categories.category_id` perhaps there is some way I can abuse or exploit the new code that splices in user-defined metadata. `resolve-fields` **should not** be resolving fields that aren't in the query -- this is a feature and not a bug. The fact that this PR doesn't change any of the middleware like `resolve-joins` to have it ignore `:metadata/dataset-metadata` when looking for joins to add is extremely concerning to me

3. I'm not convinced that tweaking the backend API endpoints to maintain frontend edit state is a good way to fix things. I think 

   > The frontend should just ignore this metadata from running the updated query. The backend will correctly handle this situation with the original edited metadata

   makes much more sense. The backend endpoints shouldn't need to have the ability to parrot back metadata, it's complicated enough already without also managing the frontend's state for it. The fact that trying to do this is causing bugs in the Query Processor is a sign that maybe this is not the best way to approach this problem -- it seems like a big hack that needs tons of band-aids to make work

4. If you **do** want to parrot back metadata, then I really don't think we should be including it in the actual query sent to the QP at all. We should just let the API endpoint code merge it in in the response metadata (this might require a bit of hacking because the responses are async). 
 
   `validate-temporal-bucketing` was failing because it didn't know that the `:metadata/dataset-metadata` is not actually part of the query. Resolving fields is only a half fix -- while `validate-temporal-bucketing` and other middleware doesn't fail anymore, all of the middleware that walks the query and transforms things is still going to transform `:metadata/dataset-metadata` because nothing knows that it should be ignored because it is metadata. Things like `resolve-joins` is going to walk it too. If the extra metadata wasn't passed to the QP then we wouldn't have to worry about that, and it would resolve the security concerns.

### TL;DR

Either have the frontend manage its own state (no backend changes), **or** have the API endpoint splice in the extra metadata without ever sending it to the QP.

**Comment by gusaiani** [comment]
@dpsutton, what are your thoughts on the above?

One of the things I'm thinking about:

> Right now when previewing the edited query it does not show the effects of the edited metadata

From what you can see of the BE, do we need it to be aware of the latest metadata for it to output the most up-to-date query results and so on?

**Comment by dpsutton** [comment]
> From what you can see of the BE, do we need it to be aware of the latest metadata for it to output the most up-to-date query results and so on?

If there are unsaved metadata edits the backend cannot incorporate those without them being sent over.

That said, I'm still a massive fan of just the FE ignoring the metadata from preview runs. Just don't overcomplicate it and _certainly_ don't throw away user edits to metadata.

I think landing that makes a lot of sense and removes an impediment to model adoption. Once that is done we can consider ways to have the preview results include the metadata edits.

For improvements in the future, I think @camsaul 's idea to merge the edits without sending them through the qp has some merit. Another option is to merge the sent metadata with the existing metadata before the qp which merges on field-refs

**Comment by gusaiani** [comment]
@dpsutton thanks.

As a next step, we will not send metadata from the FE and will  work out the metadata from cache also on the FE once the request gets a response.

**Comment by gusaiani** [comment]
As a next step, we will not send metadata from the FE and will work out the metadata from cache also on the FE once the request gets a response.
